src/server/frontend_xwayland/xwayland_wm.cpp

```diff
@@ -492,10 +518,13 @@ void mf::XWaylandWM::handle_surface_id(std::shared_ptr<XWaylandWMSurface> surfac
     uint32_t id = event->data.data32[0];
     surface->set_surface_id(id);
 
-    wl_resource *resource = wl_client_get_object(wlclient, id);
-    auto *wlsurface = resource ? WlSurface::from(resource) : nullptr;
-    if (wlsurface)
-        surface->set_surface(wlsurface);
+    run_on_wayland_dispatch([&, this]()
+        {
+            wl_resource* resource = wl_client_get_object(wlclient, id);
+            auto* wlsurface = resource ? WlSurface::from(resource) : nullptr;
+            if (wlsurface)
+                surface->set_surface(wlsurface);
+        });
```

Comment:
This could lead to excessive blocking of the X11 thread if many windows need to be looked up at once. Is there a reason we can't capture everything we need, jump into the Wayland thread with `wlc->run_on_wayland_display()` and finish up there without blocking X11?

Reply:
Hm. Is there anything part of the system that assumes that the surface_id message is handled before something else happens? If not, it would be perfectly sensible to just post this to the Wayland thread rather than waiting for it to be handled.